### DISCUSSION TITLE
Don't use loops in tests

### DISCUSSION THREAD

**Discussion by ccea-dev-team** [selenide/selenide#2536] [Q&A] [answered]
Referencing blog post here -https://selenide.org/2023/10/26/selenide-7.0.1/

Without looping what's the best way to click 50 checkboxes on a page?

![Capture](https://github.com/selenide/selenide/assets/99395943/70206935-95a0-44e7-bd59-b995b5f36d74)

**Comment by asolntsev** [comment]
I would say, it's not a good idea to click 50 checkboxes because 
1. it makes the test very slow, and
2. this is a very terrible design for end-users :)

From the test point of view, how do you know that this loop goes over exactly 50 checkboxes?
```
for (SelenideElement checkbox : $$("input")) {
  checkbox.click();
}
```

What if the locator was invalid, and there were 0 checkboxes?

**Comment by ccea-dev-team** [comment]
As dev youself you should know we get weird requests!  So for example we have an online application form where the user has to agree to maybe 20-30 statements by clicking a checkbox beside each of them.  The business people want it this way rather than a sensible way like having 1 checkbox to agree to all (maybe they think it means more chance the user reads the statements rather than clicking 1 box and going to next page).

I was looping round all the < tr > elements in the table and clicking the checkbox contained in a < td >.

**Reply by asolntsev** [reply]
@ccea-dev-team What about my second question?
> how do you know that this loop goes over exactly 50 checkboxes?
> What if the locator was invalid, and there were 0 checkboxes?

**Reply by ccea-dev-team** [reply]
>how do you know that this loop goes over exactly 50 checkboxes?

All the check boxes in the table are clicked by looping through the rows in the table.  I'm not sure I understand what you mean?  If the table has 20 rows it will have 20 checkboxes and if you loop through the table it will click the checkbox in each row. 

>What if the locator was invalid

This would be a problem for any test? Again not sure I understand what you are getting at.  I know the table name by it's id and can get the all rows using $$("#table tr)" and I know each row has a check box.

>and there were 0 checkboxes?

That would cause the test to fail which would be expected.

**Reply by ccea-dev-team** [reply]
I bit more clarification - There is a check the table row size matches the number of statements for that section of the application.  The statements are stored in a database table and I do a count of the number of statements and check that the table row size matches this number using .shouldBe(CollectionsCondition.size(statementCount)).  The statements are displayed on the screen using a < ui:repeat > element.

This test is an end to end test of the application process and I need to click all the boxes to proceed to the next step.  if you don't recommend looping, is there a better way of clicking all these check boxes?  Maybe a quick javascript way?

**Reply by briane80** [reply]
Me again but on personal account as not in work.  Mistake above - It's not using a < ui:repeat > just a h:datatable with the rows loaded from a database query getting the statements for the application section.

But regardless of how the check boxes are there, just assume there are 20 on a page that are in a table with 1 in each row and that your test needs to click all these to proceed.  We looped through this table easy enough and clicked the check boxes but after updating to version 7.0.0 I had to add .asFixedIterable to these kind of tests as the ability to loop through an element collection was removed.  The explanation of this removal was becase we shouldn't be doing this, and my question is what other way?

**Reply by asolntsev** [reply, marked as answer]
Hi @ccea-dev-team and @briane80 :)
_The first rule of GitHub Club: you do not talk about GitHub Club :)_ 

1. If you check the collection size, then it's not bad. Then you can loop. :) 
  1. 2. If you don't see the risk that the collection elements might be changed/updated/reloaded during the loop.
2. In Selenide 7.0.1, we restored the loop. Now you can remove `asFixedIterable` again.
3. Yes, much faster would be to click all 50 checkboxes with a single JS call.
4. The real reason of removing method `$$.iterator()` was that it was not clear if this methods reloads collection elements every time, or loads them only once. Each of these two approaches has pros and cons (speed vs instability). So we decided to implement both approaches and give them explicit names (`asFixedIterable` and `asDynamicIterable`). So people can decide which approach they need in every specific situation. 
Now the restored method `iterator` behaves like `asFixedIterable`, and this is not obvious for readers. That's why I don't like it.


Anyway, thank you for asking challenging questions and discussing the important decision. This is the way how open-source evolves.
